src/nbl/asset/utils/CGLSLCompiler.cpp

```diff
@@ -12,29 +12,155 @@
+
+            if (std::filesystem::exists(name) && !reqBuiltin)
+                name = std::filesystem::absolute(name);
+
+            if (_type == shaderc_include_type_relative)
+                res_str = m_defaultIncludeFinder->getIncludeRelative(relDir, _requested_source);
+            else //shaderc_include_type_standard
+                res_str = m_defaultIncludeFinder->getIncludeStandard(relDir, _requested_source);
+
+            if (!res_str.size()) {
+                const char* error_str = "Could not open file";
+                res->content_length = strlen(error_str);
+                res->content = new char[res->content_length + 1u];
+                strcpy(const_cast<char*>(res->content), error_str);
+                res->source_name_length = 0u;
+                res->source_name = "";
+            }
+            else {
+                //employ encloseWithinExtraInclGuards() in order to prevent infinite loop of (not necesarilly direct) self-inclusions while other # directives (incl guards among them) are disabled
+                IShaderCompiler::disableAllDirectivesExceptIncludes(res_str);
+                disableGlDirectives(res_str);
+                res_str = IShaderCompiler::encloseWithinExtraInclGuards(std::move(res_str), m_maxInclCnt, name.string().c_str());
+
+                res->content_length = res_str.size();
+                res->content = new char[res_str.size() + 1u];
+                strcpy(const_cast<char*>(res->content), res_str.c_str());
+                res->source_name_length = name.native().size();
+                res->source_name = new char[name.native().size() + 1u];
+                strcpy(const_cast<char*>(res->source_name), name.string().c_str());
+            }
+
+            return res;
+        }
+
+        void ReleaseInclude(shaderc_include_result* data) override
+        {
+            if (data->content_length > 0u)
+                delete[] data->content;
+            if (data->source_name_length > 0u)
+                delete[] data->source_name;
+            delete data;
+        }
+    };
+}
 
 CGLSLCompiler::CGLSLCompiler(core::smart_refctd_ptr<system::ISystem>&& system)
     : IShaderCompiler(std::move(system))
 {
 }
 
-core::smart_refctd_ptr<ICPUShader> CGLSLCompiler::compileToSPIRV(const char* code, const IShaderCompiler::SCompilerOptions& options) const
+
+
+std::string CGLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADER_STAGE& stage, const SPreprocessorOptions& preprocessOptions) const
 {
-    auto glslOptions = option_cast(options);
+    if (preprocessOptions.extraDefines.size())
+    {
+        insertExtraDefines(code, preprocessOptions.extraDefines);
+    }
+    IShaderCompiler::disableAllDirectivesExceptIncludes(code);
+    disableGlDirectives(code);
+    shaderc::Compiler comp;
+    shaderc::CompileOptions options;
+    options.SetTargetSpirv(shaderc_spirv_version_1_6);
 
-    if (!code)
+    if (preprocessOptions.includeFinder != nullptr)
```

Comment:
Have you tested what happens when you don't have an include finder?

Reply:
yes, i'll work but won't do includes